src/vs/vscode.d.ts

```diff
@@ -4587,6 +4623,110 @@ declare module 'vscode' {
 		 * Dispose and free associated resources.
 		 */
 		dispose(): void;
+
+		/**
+		 * Fires when the terminal's pty slave pseudo-device is written to. In other words, this
+		 * provides access to the raw data stream from the process running within the terminal,
+		 * including VT sequences.
+		 */
+		readonly onDidWriteData: Event<string>;
```

Comment:
@jrieken requesting an exception to the global event rule for this for performance reasons https://github.com/Microsoft/vscode/issues/67924#issuecomment-460637292

Also should we call out the performance implications in the jsdoc?

Reply:
> for this for performance reasons [#67924 (comment)](https://github.com/Microsoft/vscode/issues/67924#issuecomment-460637292)

Why? Did you measure this? What are the differences here?

Reply:
No I didn't measure, but any data going to any terminal would need to be sent over to the extension host, for any terminal. For a command that generates a lot of output that could cause the extension host to be less responsive and cause a bunch of unnecessary work (which eats at battery life as well). My thinking is that it's better to only do this for the specific terminal(s) for which it's requested, a global event would need to send it for all terminals.

Reply:
I see. Needs thinking... An exception from a rule hurts, everyone will forever point at this and do whatever they want...

Reply:
Ideas/questions from the standup

* would thing be easier if terminals are managed from the extension host so that the data is already in the right process?
* what terminal needs to be listened on? those that you create or also those that are around?
* should there be a creation/instance flag that control is a terminal sends global events or not?
* what happens to with those event that happen after creating a terminal and before attaching a listener to that terminal (is this like the linux file event dilemma?)

Reply:
> would things be easier if terminals are managed from the extension host so that the data is already in the right process?

They would still need to travel to the renderer to be displayed so not really, investigations into keeping the processes on the extension host revealed other downsides like it's harder to keep the size of the pty in sync and we would need to create out own throttling mechanism to ensure the process doesn't get too far ahead of the UI.

> what terminal needs to be listened on? those that you create or also those that are around?

Live share needs to listen to any terminal created, not just those that are created by the extension.

> should there be a creation/instantiation flag that controls if a terminal sends global events or not?

No go because it's not just for terminals created via an extension.

> what happens to with those event that happen after creating a terminal and before attaching a listener to that terminal (is this like the linux file event dilemma?)

I suspect this is not an issue right now purely because the timing to create the terminal process takes longer than the trip for the listener request. If this does end up being a problem we could keep the first x ms of data or something just in case a listener request comes in late.

---

Another option is to have global events, but similar to now, only start registering listeners and sending the data over if there is at least 1 listener to the event. Then drop the listeners when the event is disposed.

Maybe this is the best of both worlds, I just use Live Share that would touch this API so that would mean it sends events for all terminals during a live share session, and no events when the session is finished (provided the extension event listeners are disposed correctly).